# Kapitel 5 Dimensionsreducering

## Faktafrågor

### Uppgift 1. 
*Curse of dimensionality* (dimensionsförbannelsen) innebär att när antalet variabler/dimensioner i ett dataset ökar blir problmet för maskininlärningsmodellen betydligt mer komplext.
När dimensionerna blir många:
- krävs mer data för att modellen ska kunna lära sig bra,
- blir datapunkterna glesare,
- ökar risken för överanspassning (overfitting),
- kan beräkningar och modellträning ta längre tid,
- blir det svårare att visualisera och tolka datan

Ett exempel är att 100 observationer kan vara ganska mycket information om man har två variabler, men väldigt lite information om man har flera hundra variabler

### Uppgift 2.
Dimensionsreducering innebär att man minskar antalet variabler/dimensioner i ett dataset, samtidigt som man försöker behålla så mycket viktig information som möjligt.

Det görs bland annat för att:
- minska risken för overfitting,
- göra modeller snabbare,
- minska mängden data som behöver bearbetas,
- ta bort deundanta eller mindre viktiga egenskaper,
- göra data lättare att visualiser,
- ibland förbättra modellens prestanda.
Ett vanligt sätt att göra dimensionsreducering är **PCA (Principal Component Analysis)**

### Uppgift 3. 
PCA (Principal Component Analysis) fungerar i korthet ut på att hitta de riktningar i datan där spridningen (variansen) är som störst, och använda dessa för att komprimera ner dimensionerna.

Med utgångspunkt i figuren kan man förklara processen i tre steg:

1. **Hitta huvudkomponenter (vänstra diagrammet):**
Datan i det vänstra diagrammet har två ursprungliga dimensioner ($X_1$ och $X_2$). Istället för att bara titta på axlarna letar PCA efter linjer som fångar upp mest information. Den första huvudkomponenten ($C_1$, den helt dragna linjen) dras längs med den riktning där datan är mest utspridd. Den andra komponenten ($C_2$, den streckade linjen) står vinkelrätt mot den första och fångar upp återstående varians.

2. **Projicera ner på en ny axel (högra diagrammet):**
När vi vill minska dimensionerna (t.ex. från 2D till 1D) projicerar vi datapunkterna vinkelrätt ner på en av dessa linjer.

- Om vi projicerar på $C_1$ (översta grafen till höger, $Z_1$) behålls spridningen i datan väldigt väl eftersom linjen följer molnets längsta utsträckning.
- Om vi istället skulle projicera på en dålig axel (som de två undre graferna visar) pressas punkterna ihop och mycket mer information (varians) går förlorad.

3. **Resultat:**
Genom att välja den axel som har störst spridning ($C_1$) maximerar vi den information som sparas, vilket gör att vi kan reducera antalet dimensioner (från 2 till 1) med minimal informationsförlust.

### Uppgift 4. 
Kernel OCA kan utvärderas genom att udersöka hur bra den reducerade representationen fungerar för den uppgift man vill lösa.
Man kan exempelvis:
- jämföra modellens prestanda före och efter dimensionsreduceringen,
- använda de reducerade komponenterna som input till en maskininlärningsmodell och jämföra exempelvis **accuracy, precision, recall eller RMSE**, 
- undersöka hur mycket information som verkar bevaras,
- visualisera datan efter reduktionen och se om olika klasser blir tydligare separerade.

Det finns alltså inte nödvändigtvis ett enda mått som säger att kernel PCA är "bra". Man bör istället utvärdera om den nya representationen faktiskt hjälper den efterföljande ML-uppgiften.

## Resonemangsfrågor

### Uppgift 5. 
Jag håller med Kalle om att bästa möjliga prediktioner inte alltid är det enda målet. 
En modell måste bedömas utifrån både prestanda och kostnad, där tid kan vara en viktig faktor. 

En mycket komplex modell kanske ger 1 % bättre prediktioner men samtidigt kräver flera timmar att träna och flera sekunder för varje prediktion. En enklare modell kanske ger nästan lika bra resultat men tränas på några minuter och ger prediktioner på millisekunder.

I exempelvis ett system som behöver fatta besulut i realtid kan prediktionstiden vara mycket viktigare än en liten förbättring i modellens accuracy. 

ÄVen träningstiden spelar roll. Om en modell behöver tränas om ofta kan e modell som tar 10 minuter att räna vara btydligt mer praktisk än en modell som tar flera timmar.

Därför bör man försöka hitta en balans mellan prediktionskvalitet, träningstid, prediktionstid och resursanvändning.

Man kan alltså säga att Stina har rätt i att bra prediktioner är viktiga, men Kalle har rätt i att den bästa modellen inte alltid är den modell som har högst prediktionsnoggrannhet. 

### Uppgift 6. 
Efter PCA blir tolkningen av de ursprungliga variablerna svårare.

Det beror på att PCA skapar nya variabler, alltså principal components, som är kombinationer av de ursprungliga variablerna.

Exempelvis kan en principal component bestå av:

PC1 = 0,6 × längd + 0,5 × vikt − 0,2 × ålder

PC1 är alltså inte längre en variabel som har en direkt och enkel betydelse som exempelvis "vikt" eller "ålder".

Fördelarna är att vi kan reducera antalet dimensioner och behålla mycket av variationen i datan. Nackdelen är att tolkbarheten minskar, eftersom de nya komponenterna representerar kombinationer av flera ursprungliga variabler.

Därför är PCA ofta användbart när modellens prestanda och dimensionsreducering är viktigare än att kunna tolka varje enskild variabel.

## Koduppgifter

### Uppgift 7. 

In [ ]:
# Koden nedan demonstrerar hur PCA kan tillämpas
import numpy as np
from sklearn.decomposition import PCA

np.random.seed(42)
X = np.random.rand(1000, 10)
pca = PCA(n_components=0.7)
print(f"vi har gått från {X.shape[1]} dimensioner till {X.reduced.shape[1]} dimensioner genom att genomföra en dimensionsreducering.")

### Uppgift 8. 

Koden demonstrerar dimensionalitetsreduktion med hjälp av PCA (**Principal Component Analysis**), vilket är en populär teknik inom oövervakad maskininlärning för att banta ner antalet variabler (features) i ett dataset samtidigt som man behåller så mycket som möjligt av informationen (variansen).

Här är en uppdelning rad för rad av vad koden gör:
1. **Skapar ett slumpmässigt dataset:**
`X = np.random.rand(1000, 3)` skapar en matris med 1 000 rader (observationer) och 3 kolumner (features) fyllda med slumpmässiga värden mellan 0 och 1. `print(X[0:5])` visar de första 5 raderna.

2. **Reducerar dimensionerna:**
`pca = PCA(n_components=2)` förbereder en PCA-modell som är inställd på att komprimera datan från 3 dimensioner ner till 2.
`X2D = pca.fit_transform(X)` anpassar modellen till datan och omvandlar de ursprungliga 3 kolumnerna till 2 nya dimensioner (eller "huvudkomponenter"). `print(X2D[0:5])` visar hur de första 5 raderna ser ut efter nedskalningen.

3. **Återskapar datan (Inverse transform):**
`X3D_inv = pca.inverse_transform(X2D)` försöker "backa" operationen och räkna om de 2 dimensionerna till 3 dimensioner igen.

4. **Jämför originalet med den återskapade datan:**
`np.allclose(X3D_inv, X)` kollar om den återskapade matrisen är i princip identisk med originalmatrisen. Svaret här blir False, eftersom komprimeringen från 3 till 2 dimensioner innebär att en del information (varians) oundvikligen har gått förlorad på vägen.


### Uppgift 9. 

1. **Så här påverkas datan och modelleringen**
- **Dimensionalitetsreduktion:**
 Efter One-Hot-kodning av kolumner som Brand och Model har datasetet runt 52 dimensioner. Genom att köra PCA (t.ex. med kravet att behålla 95% av variansen) komprimeras antalet dimensioner ner till ca 27 komponenter

- **Förlust av ursprunglig tolkningsbarhet:**
 De ursprungliga kolumnerna (som miltal, årsmodell och hästkrafter) bakas ihop till nya abstrakta axlar (huvudkomponenter). Det blir svårare att svara på exakt hur en specifik variabel påverkar bilpriset.

- **Effekt på träffsäkerhet (beroende på modell):**
  - **Trädbaserade modeller (t.ex. Random Forest):** Presterar generellt sämre eller opåverkat när man kör PCA på tabulär data. Random Forest är skickligt på att hantera olinjära samband och tröskelvärden i originalvariablerna, och PCA kan faktiskt sudda ut en del av den informationen, vilket sänker prestandan något (i exemplet sjunker $R^2$ från ~0.987 till ~0.961).

  - **Linjära modeller (t.ex. Linear Regression):** Påverkas knappt alls i träffsäkerhet eftersom linjära samband bevaras väl, men beräkningarna blir mer komprimerade.


#### Python-kod för att köra PCA på datasetet

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.decomposition import PCA
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import r2_score, mean_squared_error

In [7]:
# 1. Ladda datasetet
df = pd.read_csv("car_price_dataset.csv", sep=';')

X = df.drop(columns=['Price'])
y = df['Price']

# Identifiera numeriska och kategoriska kolumner
numeric_cols = X.select_dtypes(include=['int64', 'float64']).columns.tolist()
categorical_cols = X.select_dtypes(include=['object']).columns.tolist()

C:\Users\lindb\AppData\Local\Temp\ipykernel_25408\4004733843.py:9: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_cols = X.select_dtypes(include=['object']).columns.tolist()


In [8]:
# 2. Dela upp datan i Train, Validation och Test (70% / 15% / 15%)
X_train_val, X_test, y_train_val, y_test = train_test_split(
    X, y, test_size=0.15, random_state=42
)
X_train, X_val, y_train, y_val = train_test_split(
    X_train_val, y_train_val, test_size=0.1765, random_state=42
)

print(f"Träningsset:    {X_train.shape[0]} rader")
print(f"Valideringsset: {X_val.shape[0]} rader")
print(f"Testset:        {X_test.shape[0]} rader\n")

Träningsset:    6999 rader
Valideringsset: 1501 rader
Testset:        1500 rader



In [9]:
# 3. Preprocessors (standard för baslinje, tät för PCA)
preprocessor_standard = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), numeric_cols),
        ('cat', OneHotEncoder(handle_unknown='ignore'), categorical_cols)
    ]
)

preprocessor_dense = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), numeric_cols),
        ('cat', OneHotEncoder(handle_unknown='ignore', sparse_output=False), categorical_cols)
    ]
)

In [10]:
# --- 4. MODELL UTAN PCA (Baslinje) ---
baseline_pipeline = Pipeline([
    ('preprocessor', preprocessor_standard),
    ('model', RandomForestRegressor(random_state=42, n_estimators=50))
])

baseline_pipeline.fit(X_train, y_train)
val_preds_base = baseline_pipeline.predict(X_val)

r2_base = r2_score(y_val, val_preds_base)
rmse_base = np.sqrt(mean_squared_error(y_val, val_preds_base))

In [11]:
# --- 5. MODELL MED PCA ---
pca_pipeline = Pipeline([
    ('preprocessor', preprocessor_dense),
    ('pca', PCA(n_components=0.95, random_state=42)),
    ('model', RandomForestRegressor(random_state=42, n_estimators=50))
])

pca_pipeline.fit(X_train, y_train)
val_preds_pca = pca_pipeline.predict(X_val)

r2_pca = r2_score(y_val, val_preds_pca)
rmse_pca = np.sqrt(mean_squared_error(y_val, val_preds_pca))
n_components = pca_pipeline.named_steps['pca'].n_components_

In [12]:
# --- 6. UTVÄRDERING OCH JÄMFÖRELSE PÅ VAL-SET ---
print("========================================")
print("       MODELLJÄMFÖRELSE (BILPRISER)     ")
print("========================================")
print(f"Utan PCA -> R2: {r2_base:.4f} | RMSE: {rmse_base:.2f}")
print(f"Med PCA  -> R2: {r2_pca:.4f} | RMSE: {rmse_pca:.2f} (Komponenter: {n_components})")

       MODELLJÄMFÖRELSE (BILPRISER)     
Utan PCA -> R2: 0.9865 | RMSE: 364.31
Med PCA  -> R2: 0.9697 | RMSE: 546.20 (Komponenter: 27)


========================================
       MODELLJÄMFÖRELSE (BILPRISER)     
========================================
Utan PCA -> R2: 0.9865 | RMSE: 364.31
Med PCA  -> R2: 0.9697 | RMSE: 546.20 (Komponenter: 27)

Som vi kan se så får vi sämre Resultat med PCA i detta fallet. 